### PR TITLE
feat(tm2/gnovm): add msg_idx in event to identify

### LINKED ISSUE #2031: [tm2/gnovm] Identify each msg's event
## Description

While fixing #2028 in #2030, I found out that all of events in single transactions are just being append into single object.

For example if we execute multi-msg tx
```javascript
await adena.DoContract({
    messages: [{
        "type": "/vm.m_call",
        "value": {
          "caller": "g1jg8mtutu9khhfwc4nxmuhcpftf0pajdhfvsqf5",
          "send": "",
          "pkg_path": "gno.land/r/demo/event",
          "func": "Hello",
          "args": []
        }
      },{
        "type": "/vm.m_call",
        "value": {
          "caller": "g1jg8mtutu9khhfwc4nxmuhcpftf0pajdhfvsqf5",
          "send": "",
          "pkg_path": "gno.land/r/demo/event",
          "func": "Other",
          "args": []
        }
      }],
    gasFee: 1, 
    gasWanted: 2000000
  });
```

Event gets like this
```json
"Events": [
    {
      "@type": "/tm.gnoEvent",
      "pkg_path": "gno.land/r/demo/event2",
      "type": "t",
      "func": "inner",
      "attrs": [
        {
          "key": "k",
          "value": "v"
        }
      ]
    },
    {
      "@type": "/tm.gnoEvent",
      "pkg_path": "gno.land/r/demo/event2",
      "type": "t",
      "func": "Hello",
      "attrs": [
        {
          "key": "k",
          "value": "v"
        }
      ]
    },
    {
      "@type": "/tm.gnoEvent",
      "pkg_path": "gno.land/r/demo/event",
      "type": "t",
      "func": "inner",
      "attrs": [
        {
          "key": "k",
          "value": "v"
        }
      ]
    },
    {
      "@type": "/tm.gnoEvent",
      "pkg_path": "gno.land/r/demo/event",
      "type": "t",
      "func": "Hello",
      "attrs": [
        {
          "key": "k",
          "value": "v"
        }
      ]
    },
    {
      "@type": "/tm.gnoEvent",
      "pkg_path": "gno.land/r/demo/event",
      "type": "other",
      "func": "Other",
      "attrs": [
        {
          "key": "k",
          "value": "v"
        }
      ]
    }
  ],
```

With current event struct, as you can see we don't know which call made which event(s).
IMHO, we need identify (or divide) each function's event

1. first group of event
```javascript
{
  "@type": "/tm.gnoEvent",
  "pkg_path": "gno.land/r/demo/event2",
  "type": "t",
  "func": "inner",
  "attrs": [
    {
      "key": "k",
      "value": "v"
    }
  ]
},
{
  "@type": "/tm.gnoEvent",
  "pkg_path": "gno.land/r/demo/event2",
  "type": "t",
  "func": "Hello",
  "attrs": [
    {
      "key": "k",
      "value": "v"
    }
  ]
},
{
  "@type": "/tm.gnoEvent",
  "pkg_path": "gno.land/r/demo/event",
  "type": "t",
  "func": "inner",
  "attrs": [
    {
      "key": "k",
      "value": "v"
    }
  ]
},
{
  "@type": "/tm.gnoEvent",
  "pkg_path": "gno.land/r/demo/event",
  "type": "t",
  "func": "Hello",
  "attrs": [
    {
      "key": "k",
      "value": "v"
    }
  ]
},
```

2. second group of event
```javascript
{
  "@type": "/tm.gnoEvent",
  "pkg_path": "gno.land/r/demo/event",
  "type": "other",
  "func": "Other",
  "attrs": [
    {
      "key": "k",
      "value": "v"
    }
  ]
}
```

### PR DESCRIPTION
**Pull request by r3v4s**
Closes #2031 

As describe in above issue, this pr adds `msg_idx` field in ctx.Event struct to identify which event was emitted by which msg.

FYI, this pr also includes another bug fix in #2030. It can be close if we're willing to merge this one

<details><summary>Contributors' checklist...</summary>

- [x] Added new tests, or not needed, or not feasible
- [x] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

**Comment by moul** [comment]
Why do you need an index when the events are delivered in an ordered slice?

Your PR only shows `msg_idx=0`, which seems incorrect.

**Comment by r3v4s** [comment]
> Why do you need an index when the events are delivered in an ordered slice?
> 
> Your PR only shows `msg_idx=0`, which seems incorrect.

As describe in #2031, current event output is just list of events without which msg emitted event if tx inlcudes 2 or more msgs.

I'll add test cases that runs 2 messages

File: gno.land/pkg/gnoclient/integration_test.go [hunk 1]
```diff
@@ -113,6 +114,64 @@ func TestCallMultiple_Integration(t *testing.T) {
 	assert.Equal(t, expected, got)
 }
 
+func TestCallMultipleWithEvent_Integration(t *testing.T) {
```
Comment (ajnavarro) [inline gno.land/pkg/gnoclient/integration_test.go:117]: Could you test generating several events between different blocks and different transactions? thanks

File: gnovm/stdlibs/std/emit_event.go [hunk 2]
```diff
@@ -46,14 +47,20 @@ func attrKeysAndValues(attrs []string) ([]gnoEventAttribute, error) {
...
+	Type       string              `json:"type"`
 	Attributes []gnoEventAttribute `json:"attrs"`
 }
 
 func (e gnoEvent) AssertABCIEvent() {}
 
+func (e gnoEvent) SetMsgIdx(msgIdx int) interface{} {
+	e.MsgIdx = msgIdx
+	return e
+}
```
Comment (ajnavarro) [inline gnovm/stdlibs/std/emit_event.go:62]: Why this instead of `e.MsgIdx = msgIdx` when needed?

File: tm2/pkg/bft/abci/types/types.go [hunk 3]
```diff
@@ -206,6 +206,7 @@ type Error interface {
 
 type Event interface {
 	AssertABCIEvent()
+	SetMsgIdx(int) interface{}
```
Comment (ajnavarro) [inline tm2/pkg/bft/abci/types/types.go:209]: If you want to force that an Event needs an ID, I would recommend adding the method providing that ID instead of the provided Setter:

```suggestion
	ID() []byte
```

If the ID is expected to be sequential:

```suggestion
	Seq() uint64
```

**Review by ajnavarro — CHANGES_REQUESTED** [review]

File: gnovm/stdlibs/std/emit_event.go (hunk 2 above)
Comment (r3v4s) [reply gnovm/stdlibs/std/emit_event.go:62]: If what you're suggesting is `e.MsgIdx = msgIdx` without ` return e`, vm was panicking if function is like `func (e *gnoEvnet)`.

File: tm2/pkg/bft/abci/types/types.go (hunk 3 above)
Comment (r3v4s) [reply tm2/pkg/bft/abci/types/types.go:209]: Looking Good!

File: gno.land/pkg/gnoclient/integration_test.go (hunk 1 above)
Comment (r3v4s) [reply gno.land/pkg/gnoclient/integration_test.go:117]: I'll add test that tests `multi tx + mutli msg`.

But I don't think testing events between different block is necessary test, different block really has nothing to do with `multi-msg`

File: tm2/pkg/bft/abci/types/types.go (hunk 3 above)
Comment (r3v4s) [reply tm2/pkg/bft/abci/types/types.go:209]: Reverted.
Event needs not only ID, but also setter for setting msg_idx.

Currently, event gets emitted from here
https://github.com/gnolang/gno/blob/d7c04bee3c90718bf8b82a00399f6aa387e82376/gnovm/stdlibs/std/emit_event.go#L14-L31
At this point, there is no way for us to determine message index if current request was made by multi-msg tx.

However `baseapp.go` does know full list of messages so, we can set message index from there.
https://github.com/gnolang/gno/blob/dabf13c705b4ac13ffd25d55a9a4a2107e335555/tm2/pkg/sdk/baseapp.go#L634-L666

File: gno.land/pkg/gnoclient/client_txs.go [hunk 4]
```diff
@@ -114,6 +114,62 @@
 	return c.signAndBroadcastTxCommit(tx, cfg.AccountNumber, cfg.SequenceNumber)
 }
 
+// CallTxSync executes one or more MsgCall calls on the blockchain synchronously
+func (c *Client) CallTxSync(cfg BaseTxCfg, msgs ...MsgCall) (*ctypes.ResultBroadcastTx, error) {
```
Comment (zivkovicmilos) [inline gno.land/pkg/gnoclient/client_txs.go:118]: We should leave this for a separate PR 🙏

File: tm2/pkg/sdk/baseapp.go [hunk 5]
```diff
@@ -652,25 +652,34 @@ func (app *BaseApp) runMsgs(ctx Context, msgs []Msg, mode RunTxMode) (result Res
...
 		data = append(data, msgResult.Data...)
-		events = append(events, msgResult.Events...)
-		defer func() {
-			events = append(events, ctx.EventLogger().Events()...)
-			result.Events = events
-		}()
-		// TODO append msgevent from ctx. XXX XXX
+
+		// each msgs' event
+		var msgEvent []abci.Event
```
Comment (zivkovicmilos) [inline tm2/pkg/sdk/baseapp.go:657]: You know at this point in time how many events you'll have, so this slice should be initialized with a capacity:
```go
msgEvent := make([]abci.Event, 0, len(msgCtxEvents))
```

File: gnovm/stdlibs/std/emit_event.go [hunk 6]
```diff
@@ -46,14 +47,20 @@ func attrKeysAndValues(attrs []string) ([]gnoEventAttribute, error) {
...
+	MsgIdx     int                 `json:"msg_idx"`
 	PkgPath    string              `json:"pkg_path"`
 	Func       string              `json:"func"`
+	Type       string              `json:"type"`
 	Attributes []gnoEventAttribute `json:"attrs"`
 }
 
 func (e gnoEvent) AssertABCIEvent() {}
 
+func (e gnoEvent) SetMsgIdx(msgIdx int) interface{} {
```
Comment (zivkovicmilos) [inline gnovm/stdlibs/std/emit_event.go:59]: Why do you need to return the message event from this method?

File: tm2/pkg/sdk/baseapp.go [hunk 7]
```diff
@@ -652,25 +652,34 @@ func (app *BaseApp) runMsgs(ctx Context, msgs []Msg, mode RunTxMode) (result Res
...
+
+		// handle each msg's ctx event
+		msgCtxEvents := ctx.EventLogger().Events()
+		for _, msgCtxEvent := range msgCtxEvents {
```
Comment (zivkovicmilos) [inline tm2/pkg/sdk/baseapp.go:661]: I'm not sure what kind of performance overhead this has, but it can't be minimal

File: tm2/pkg/bft/abci/types/types.go (hunk 3 above)
Comment (zivkovicmilos) [inline tm2/pkg/bft/abci/types/types.go:209]: I'm gonna suggest something radical, but hear me out

We don't actually need this information in the event. 
Why would it matter which event belongs to which message?

How can you guarantee that the node doesn't shuffle the messages before executing them? (it doesn't now, but I see no reason why it might not)

I'm not in favor of this idea because it:
- introduces more complexity on the already bloated event API
- it's just a QoL feature that doesn't have any real functional value, since events are async and we should be concerned with their content, not their origin

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
Please check my comments, I don't think we need this functionality in the events 🙁

File: gno.land/pkg/gnoclient/client_txs.go (hunk 4 above)
Comment (r3v4s) [reply gno.land/pkg/gnoclient/client_txs.go:118]: will do

File: tm2/pkg/bft/abci/types/types.go (hunk 3 above)
Comment (r3v4s) [reply tm2/pkg/bft/abci/types/types.go:209]: > We don't actually need this information in the event.
Why would it matter which event belongs to which message?

- Let't say... single block has N txs, and each txs has N2 messages and each message has N3 emitted events.

- If all of N2 messages are same function with just different input params, this is situation where I use `msg_idx` to identify it.

- One of gnoswap(+interface) functionality is removing all of your positions in one shot, by calling `RemovePosition(positionId)` several time(≈several msgs). 

- If that `RemovePosition` emits few event with some underlying infos. At this point, we need to know which events was came from which message (i.e which position emitted which events)

Comment (r3v4s) [reply tm2/pkg/bft/abci/types/types.go:209]: > How can you guarantee that the node doesn't shuffle the messages before executing them? (it doesn't now, but I see no reason why it might not)
- I really didn't think about this :(

> introduces more complexity on the already bloated event API
- Agreed

Comment (zivkovicmilos) [reply tm2/pkg/bft/abci/types/types.go:209]: > If all of N2 messages are same function with just different input params, this is situation where I use msg_idx to identify it.

Yes, but why? Isn't it enough that you can parse `RemovePosition` events on the client, and use data from the event content to figure out what to do?

Maybe I'm missing something, but tying events to their respective messages doesn't solve any issue -- you should be able to adequately parse them client-side (reactive, async) and execute logic based on their appearance. This is the foundation of [event driven design](https://en.wikipedia.org/wiki/Event-driven_architecture)

@ajnavarro @gfanton maybe I am missing something?

**Comment by zivkovicmilos** [comment]
After our PR discussions, and discussions with @r3v4s, we've decided to close this PR. The bulk of the reasoning is outlined here:
https://github.com/gnolang/gno/pull/2061#discussion_r1602884318
